oxidation/libparsec/crates/platform-async/examples/eyeballs.rs

```diff
@@ -0,0 +1,137 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BSLv1.1 (eventually AGPLv3) 2016-2021 Scille SAS
+#![cfg(not(target_arch = "wasm32"))]
+
+use libparsec_platform_async::{
+    platform::{spawn, Task},
+    task::Taskable,
+};
+use std::{
+    sync::{Arc, Mutex},
+    time::{Duration, Instant},
+};
+use tokio::sync::{mpsc, Notify};
+
+const TIMEOUT: u64 = 250;
+
+#[derive(Clone, Copy, Debug)]
+enum Mode {
+    Fail(Duration),
+    Success(Duration),
+}
+
+#[tokio::main]
+async fn main() -> Result<(), ()> {
+    simple_logger::init_with_level(log::Level::Debug).expect("cannot initialise simple logger");
+
+    let targets = vec![
+        Mode::Fail(Duration::from_millis(2300)),
+        Mode::Fail(Duration::from_millis(300)),
+        Mode::Success(Duration::from_millis(300)),
+        Mode::Success(Duration::from_millis(100)),
+    ];
+
+    log::debug!("starting with {} targets", targets.len());
+    let start_time = Instant::now();
+    assert_eq!(Ok(2), dbg!(eyeballs(targets).await));
+    let elapsed = start_time.elapsed();
+    log::info!("elapsed time {}ms", elapsed.as_millis());
+    assert!(elapsed < Duration::from_millis(900));
+    Ok(())
+}
+
+async fn eyeballs(targets: Vec<Mode>) -> Result<usize, ()> {
+    let failed_attempt = targets
+        .iter()
+        .map(|_| Arc::new(Notify::new()))
+        .collect::<Vec<_>>();
+    let targets = Arc::new(targets);
+    let join_set = Vec::with_capacity(targets.len());
+    let join_set = Arc::new(Mutex::new(join_set));
+    let (tx, mut rx) = mpsc::channel::<usize>(targets.len());
+
+    spawn_attempt(0, targets, failed_attempt, join_set.clone(), tx);
+    let result = dbg!(rx.recv().await);
+    for task in join_set.lock().unwrap().iter() {
+        task.cancel();
+    }
+    result.ok_or(())
```

Comment:
What's the point to transform the option ?

Reply:
none

Reply:
changed in bb524deee